test/tests/functional/pbs_pbsnodes_attributes.py

```diff
@@ -0,0 +1,115 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2018 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class Testpbsnodes_attributes(TestFunctional):
+
+    def Test_pbsnodes(self, user):
+        """
+        Function to validate pbsnodes attribute on user basis
+        """
+
+        attr_dict = {}
+        host_short_name = self.mom.shortname
+        host_name = self.mom.hostname
+        pbs_version = self.server.pbs_version
+
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'pbsnodes -a')
+        if user == 'root':
+            ret = self.du.run_cmd(self.server.hostname, command, sudo=True)
+        else:
+            ret = self.du.run_cmd(self.server.hostname, command, runas=user)
+
+        #converting pbsnodes -a output into dictionary for validation
+        attr_list = ret['out']
+        list_len = len(attr_list) - 1
+        attr_list = filter(None, attr_list)
+        for i in range(1, list_len):
```

Comment:
Suggestion: instead of going over the list, creating a dictionary from it, and then verifying each attribute explicitly, do the following:
1) Create a dictionary of attributes and their default, expected values.
2) Loop over the list ret['out'], look up the attribute name in the dictionary above and match the attribute value with the one in the dictionary.

That's it, this will reduce the size of your function to less than half, and you'll also be able to validate that no additional attributes were set on the vnode.

Reply:
it  is a good suggestion. this can be done on certain attributes that do not have more than one value. for the ones that has more than one or vary based on the platform will still need to be on case by case basis.